evm/src/arithmetic/shift.rs

```diff
@@ -0,0 +1,394 @@
+
+    let input0 = read_value_i64_limbs(lv, INPUT_REGISTER_1); // input
+    let input1 = read_value_i64_limbs(lv, INPUT_REGISTER_2); // 1 << shift
+
+    if is_shl {
+        // If the operation is SHL, we compute `input * shifted_displacement`.
+        const MASK: i64 = (1i64 << LIMB_BITS) - 1i64;
+
+        // Input and output have 16-bit limbs
+        let mut output_limbs = [0i64; N_LIMBS];
+
+        // Column-wise pen-and-paper long multiplication on 16-bit limbs.
+        // First calculate the coefficients of a(x)*b(x) (in unreduced_prod),
+        // then do carry propagation to obtain C = c(β) = a(β)*b(β).
+        let mut cy = 0i64;
+        let mut unreduced_prod = pol_mul_lo(input0, input1);
+        for col in 0..N_LIMBS {
+            let t = unreduced_prod[col] + cy;
+            cy = t >> LIMB_BITS;
+            output_limbs[col] = t & MASK;
+        }
+        // In principle, the last cy could be dropped because this is
+        // multiplication modulo 2^256. However, we need it below for
+        // aux_limbs to handle the fact that unreduced_prod will
+        // inevitably contain one digit's worth that is > 2^256.
+
+        pol_sub_assign(&mut unreduced_prod, &output_limbs);
+
+        let mut aux_limbs = pol_remove_root_2exp::<LIMB_BITS, _, N_LIMBS>(unreduced_prod);
+        aux_limbs[N_LIMBS - 1] = -cy;
+
+        for c in aux_limbs.iter_mut() {
+            // we store the unsigned offset value c + 2^20
+            *c += AUX_COEFF_ABS_MAX;
+        }
+
+        debug_assert!(aux_limbs.iter().all(|&c| c.abs() <= 2 * AUX_COEFF_ABS_MAX));
+
+        lv[MUL_AUX_INPUT_LO].copy_from_slice(&aux_limbs.map(|c| F::from_canonical_u16(c as u16)));
+        lv[MUL_AUX_INPUT_HI]
+            .copy_from_slice(&aux_limbs.map(|c| F::from_canonical_u16((c >> 16) as u16)));
+    } else {
+        // If the operation is SHR, we compute: `input / shifted_displacement` if `shifted_displacement == 0`
+        // otherwise, the output is 0.
+        let input_limbs = read_value_i64_limbs::<N_LIMBS, _>(lv, INPUT_REGISTER_1);
+        let pol_input = pol_extend(input_limbs);
+        let (out, quo_input) =
+            generate_modular_op(lv, nv, columns::IS_SHL, pol_input, INPUT_REGISTER_2);
+        debug_assert!(
+            &quo_input[N_LIMBS..].iter().all(|&x| x == F::ZERO),
+            "expected top half of quo_input to be zero"
+        );
+
+        // Initialise whole (double) register to zero; the low half will
+        // be overwritten via lv[AUX_INPUT_REGISTER] below.
+        for i in MODULAR_QUO_INPUT {
+            lv[i] = F::ZERO;
+        }
+
+        lv[AUX_INPUT_REGISTER_0].copy_from_slice(&out);
+    }
+}
+
+/// Evaluates the constraints for an SHL opcode.
+/// The logic is very similar to the one for MUL. The only difference is that
+/// the inputs are in `INPUT_REGISTER_1`  and `INPUT_REGISTER_2` instead of
+/// `INPUT_REGISTER_0` and `INPUT_REGISTER_1`.
+pub fn eval_packed_shl<P: PackedField>(
```

Comment:
Again, I think it's preferable and easy for the code to be shared between here and `mul.rs` rather than copy-pasting it. Same applies to all the other functions from here on.